src/raw_client.rs

```diff
@@ -668,32 +667,32 @@ impl<T: Read + Write> ElectrumApi for RawClient<T> {
 
         self.increment_calls();
 
-        while !missing_responses.is_empty() {
-            let req_id = *missing_responses.iter().next().unwrap();
-            let resp = match self.recv(&receiver, req_id) {
-                Ok(resp) => resp,
+        let mut answers = BTreeMap::new();
+        for req_id in missing_responses {
+            match self.recv(&receiver, req_id) {
                 Err(e) => {
                     // In case of error our sender could still be left in the map, depending on where
                     // the error happened. Just in case, try to remove it here
                     warn!("got error for req_id {}: {:?}", req_id, e);
                     warn!("removing all waiting req of this batch");
-                    let mut guard = self.waiting_map.lock().unwrap();
-                    for req_id in missing_responses.iter() {
-                        guard.remove(&req_id);
-                    }
+                    let mut guard = self
+                        .waiting_map
+                        .lock()
+                        .map_err(|_| Error::CouldntLockReader)?;
+                    guard.remove(&req_id);
                     return Err(e);
                 }
-            };
-            let resp_id = resp["id"].as_u64().unwrap() as usize;
-
-            missing_responses.remove(&resp_id);
-            answer.push(resp);
+                Ok(resp) => {
+                    let mut res = resp
+                        .get("result")
+                        .cloned()
+                        .ok_or(Error::InvalidResponse(resp))?;
+                    answers.insert(req_id, res.take());
+                }
+            }
         }
 
-        answer.sort_by(|a, b| a["id"].as_u64().partial_cmp(&b["id"].as_u64()).unwrap());
-        let answer = answer.into_iter().map(|mut x| x["result"].take()).collect();
-
-        Ok(answer)
+        Ok(answers.into_iter().map(|(_, item)| item).collect())
```

Comment:
I think you are missing the part where you remove the `id` of the response we've just received from `missing_responses`.